File: packages/lmon-python/lmon/macro.py
import os
import uuid
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Optional

from .errors import LMONMacroError


@dataclass
class MacroDefinition:
    body: str
    params: Optional[list[str]]
    source_line: int


MacroContext = dict[str, MacroDefinition]


@dataclass
class ExpandOptions:
    initial_context: Optional[MacroContext] = None
    strict: bool = True
    max_depth: int = 16


class ExprTokenType(Enum):
    NUMBER = "NUMBER"
    PLUS = "PLUS"
    MINUS = "MINUS"
    STAR = "STAR"
    SLASH = "SLASH"
    PERCENT = "PERCENT"
    LPAREN = "LPAREN"
    RPAREN = "RPAREN"
    EOF = "EOF"


@dataclass
class ExprToken:
    type: ExprTokenType
    value: str


class ExprEvaluator:
    def __init__(self) -> None:
        self.tokens: list[ExprToken] = []
        self.pos = 0

    def evaluate(self, expr: str, line: int, col: int) -> float:
        self.tokens = self.lex_expr(expr, line, col)
        self.pos = 0
        result = self.parse_add_expr(line, col)
        if self.current().type != ExprTokenType.EOF:
            raise LMONMacroError(
                f"Unexpected token in expression: '{self.current().value}'",
                line,
                col,
            )
        return result

    def lex_expr(self, expr: str, line: int, col: int) -> list[ExprToken]:
        tokens: list[ExprToken] = []
        i = 0

        while i < len(expr):
            ch = expr[i]

            if ch.isspace():
                i += 1
                continue

            if ch == "+":
                tokens.append(ExprToken(ExprTokenType.PLUS, "+"))
                i += 1
            elif ch == "-":
                tokens.append(ExprToken(ExprTokenType.MINUS, "-"))
                i += 1
            elif ch == "*":
                tokens.append(ExprToken(ExprTokenType.STAR, "*"))
                i += 1
            elif ch == "/":
                tokens.append(ExprToken(ExprTokenType.SLASH, "/"))
                i += 1
            elif ch == "%":
                tokens.append(ExprToken(ExprTokenType.PERCENT, "%"))
                i += 1
            elif ch == "(":
                tokens.append(ExprToken(ExprTokenType.LPAREN, "("))
                i += 1
            elif ch == ")":
                tokens.append(ExprToken(ExprTokenType.RPAREN, ")"))
                i += 1
            elif ch.isdigit() or (ch == "." and i + 1 < len(expr) and expr[i + 1].isdigit()):
                num_str = ""
                while i < len(expr) and (expr[i].isdigit() or expr[i] == "."):
                    num_str += expr[i]
                    i += 1
                tokens.append(ExprToken(ExprTokenType.NUMBER, num_str))
            else:
                raise LMONMacroError(f"Unexpected character in expression: '{ch}'", line, col)

        tokens.append(ExprToken(ExprTokenType.EOF, ""))
        return tokens

    def current(self) -> ExprToken:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return ExprToken(ExprTokenType.EOF, "")

    def advance(self) -> ExprToken:
        token = self.current()
        self.pos += 1
        return token

    def expect(self, token_type: ExprTokenType, line: int, col: int) -> ExprToken:
        token = self.current()
        if token.type != token_type:
            raise LMONMacroError(f"Expected {token_type} but got {token.type}", line, col)
        return self.advance()

    def parse_add_expr(self, line: int, col: int) -> float:
        left = self.parse_mul_expr(line, col)

        while self.current().type in (ExprTokenType.PLUS, ExprTokenType.MINUS):
            op = self.advance()
            right = self.parse_mul_expr(line, col)
            left = left + right if op.type == ExprTokenType.PLUS else left - right

        return left

    def parse_mul_expr(self, line: int, col: int) -> float:
        left = self.parse_unary(line, col)

        while self.current().type in (ExprTokenType.STAR, ExprTokenType.SLASH, ExprTokenType.PERCENT):
            op = self.advance()
            right = self.parse_unary(line, col)

            if op.type == ExprTokenType.STAR:
                left = left * right
            elif op.type == ExprTokenType.SLASH:
                if right == 0:
                    raise LMONMacroError("Division by zero in expression", line, col)
                left = left / right
            else:
                if right == 0:
                    raise LMONMacroError("Modulo by zero in expression", line, col)
                left = left % right

        return left

    def parse_unary(self, line: int, col: int) -> float:
        if self.current().type == ExprTokenType.MINUS:
            self.advance()
            return -self.parse_unary(line, col)
        return self.parse_primary(line, col)

    def parse_primary(self, line: int, col: int) -> float:
        if self.current().type == ExprTokenType.NUMBER:
            return float(self.advance().value)

        if self.current().type == ExprTokenType.LPAREN:
            self.advance()
            val = self.parse_add_expr(line, col)
            self.expect(ExprTokenType.RPAREN, line, col)
            return val

        raise LMONMacroError(
            f"Expected number or '(' in expression, got {self.current().type}",
            line,
            col,
        )


def expand(input_text: str, options: Optional[ExpandOptions] = None) -> str:
    if options is None:
        options = ExpandOptions()

    spec_macros = build_spec_macros()
    context: MacroContext = {**spec_macros, **(options.initial_context or {})}
    strict = options.strict
    max_depth = options.max_depth

    lines = input_text.split("\n")
    output_lines: list[str] = []
    trailing_newline = input_text.endswith("\n")

    for line_num in range(1, len(lines) + 1):
        line = lines[line_num - 1] if line_num <= len(lines) else ""
        import re

        def_match = re.fullmatch(
            r'^\s*%([a-zA-Z_][a-zA-Z0-9_]*)(?:\(([^)]*)\))?\s*=\s*"((?:[^"\\]|\\.)*)"[\t ]*$',
            line,
        )

        if def_match:
            name = def_match.group(1)
            param_str = def_match.group(2) or ""
            raw_body = def_match.group(3) or ""

            params: Optional[list[str]] = None
            if param_str.strip():
                params = [p.strip() for p in param_str.split(",") if p.strip()]

            body = decode_escapes(raw_body)

            context[name] = MacroDefinition(
                body=body,
                params=params,
                source_line=line_num,
            )
        else:
            expanded = expand_line(line, context, strict, max_depth, line_num, 0)
            output_lines.append(expanded)

    result = "\n".join(output_lines)
    if trailing_newline:
        result += "\n"
    return result


def decode_escapes(s: str) -> str:
    return (
        s.replace('\\"', '"')
        .replace("\\\\", "\\")
        .replace("\\n", "\n")
        .replace("\\t", "\t")
    )


def build_spec_macros() -> MacroContext:
    now = datetime.now()

    date_str = now.strftime("%Y-%m-%d")
    time_str = now.strftime("%H:%M:%S")
    datetime_str = now.isoformat() + "Z"

    days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
    day_str = days[now.weekday()]

    timestamp = str(int(now.timestamp()))
    uuid_str = str(uuid.uuid4())

    return {
        "_DATE_STR": MacroDefinition(body=date_str, params=None, source_line=0),
        "_TIMESTAMP": MacroDefinition(body=timestamp, params=None, source_line=0),
        "_DATETIME_STR": MacroDefinition(body=datetime_str, params=None, source_line=0),
        "_DAY_STR": MacroDefinition(body=day_str, params=None, source_line=0),
        "_TIME_STR": MacroDefinition(body=time_str, params=None, source_line=0),
        "_UUID": MacroDefinition(body=uuid_str, params=None, source_line=0),
        "_ENV": MacroDefinition(body="", params=["VAR"], source_line=0),
    }


def expand_line(
    line: str,
    context: MacroContext,
    strict: bool,
    max_depth: int,
    line_num: int,
    depth: int,
) -> str:
    if depth > max_depth:
        raise LMONMacroError("Macro expansion depth exceeded", line_num, 1)

    result = ""
    i = 0

    while i < len(line):
        if line[i] == "%":
            if i + 1 < len(line) and line[i + 1] == "{":
                close_idx = line.find("}", i + 2)
                if close_idx == -1:
                    raise LMONMacroError(
                        "Unclosed expression macro %{...}",
                        line_num,
                        i + 1,
                    )

                expr_str = line[i + 2 : close_idx]
                expr_str = expand_line(expr_str, context, strict, max_depth, line_num, depth + 1)

                evaluator = ExprEvaluator()
                expr_result = evaluator.evaluate(expr_str, line_num, i + 1)
                stringified = str(int(expr_result)) if expr_result == int(expr_result) else str(expr_result)

                result += stringified
                i = close_idx + 1
            else:
                import re

                name_match = re.match(r"^([a-zA-Z_][a-zA-Z0-9_]*)", line[i + 1 :])
                if not name_match:
                    result += "%"
                    i += 1
                    continue

                name = name_match.group(1)
                next_idx = i + 1 + len(name)
                args: Optional[list[str]] = None

                if next_idx < len(line) and line[next_idx] == "(":
                    close_paren_idx = line.find(")", next_idx)
                    if close_paren_idx == -1:
                        raise LMONMacroError(
                            f"Unclosed parameter list for macro '{name}'",
                            line_num,
                            next_idx + 1,
                        )

                    arg_str = line[next_idx + 1 : close_paren_idx]
                    args = [a.strip() for a in arg_str.split(",")] if arg_str.strip() else []
                    next_idx = close_paren_idx + 1

                def_macro = context.get(name)

                if not def_macro:
                    if strict:
                        raise LMONMacroError(
                            f"Undefined macro '{name}'",
                            line_num,
                            i + 1,
                            name,
                        )
                    result += f"%{name}" + (f"({','.join(args)})" if args else "")
                    i = next_idx
                    continue

                if def_macro.params is None and args and len(args) > 0:
                    raise LMONMacroError(
                        f"Macro '{name}' takes no parameters, but {len(args)} provided",
                        line_num,
                        i + 1,
                        name,
                    )

                if def_macro.params is not None and (
                    not args or len(args) != len(def_macro.params)
                ):
                    expected = len(def_macro.params)
                    got = len(args) if args else 0
                    raise LMONMacroError(
                        f"Macro '{name}' expects {expected} arguments, got {got}",
                        line_num,
                        i + 1,
                        name,
                    )

                if name == "_ENV" and args and len(args) > 0:
                    env_var_name = args[0]
                    env_value = os.environ.get(env_var_name, "")
                    substituted = env_value
                else:
                    body = def_macro.body

                    if def_macro.params and args:
                        for j, param in enumerate(def_macro.params):
                            placeholder = f"{{{param}}}"
                            body = body.replace(placeholder, args[j])

                    substituted = expand_line(body, context, strict, max_depth, line_num, depth + 1)

                result += substituted
                i = next_idx
        else:
            result += line[i]
            i += 1

    return result

File: packages/lmon/src/macro.ts
import { LMONMacroError } from './errors';

export interface MacroDefinition {
  body: string;
  params: string[] | null;
  sourceLine: number;
}

export type MacroContext = Map<string, MacroDefinition>;

export interface ExpandOptions {
  initialContext?: MacroContext;
  strict?: boolean;
  maxDepth?: number;
}

enum ExprTokenType {
  NUMBER = 'NUMBER',
  PLUS = 'PLUS',
  MINUS = 'MINUS',
  STAR = 'STAR',
  SLASH = 'SLASH',
  PERCENT = 'PERCENT',
  LPAREN = 'LPAREN',
  RPAREN = 'RPAREN',
  EOF = 'EOF',
}

interface ExprToken {
  type: ExprTokenType;
  value: string;
}

class ExprEvaluator {
  private tokens: ExprToken[] = [];
  private pos = 0;

  evaluate(expr: string, line: number, col: number): number {
    this.tokens = this.lexExpr(expr, line, col);
    this.pos = 0;
    const result = this.parseAddExpr(line, col);
    if (this.current().type !== ExprTokenType.EOF) {
      throw new LMONMacroError(
        `Unexpected token in expression: '${this.current().value}'`,
        line,
        col,
      );
    }
    return result;
  }

  private lexExpr(expr: string, line: number, col: number): ExprToken[] {
    const tokens: ExprToken[] = [];
    let i = 0;

    while (i < expr.length) {
      const ch = expr[i];
      if (ch === undefined) break;

      if (/\s/.test(ch)) {
        i++;
        continue;
      }

      if (ch === '+') {
        tokens.push({ type: ExprTokenType.PLUS, value: '+' });
        i++;
      } else if (ch === '-') {
        tokens.push({ type: ExprTokenType.MINUS, value: '-' });
        i++;
      } else if (ch === '*') {
        tokens.push({ type: ExprTokenType.STAR, value: '*' });
        i++;
      } else if (ch === '/') {
        tokens.push({ type: ExprTokenType.SLASH, value: '/' });
        i++;
      } else if (ch === '%') {
        tokens.push({ type: ExprTokenType.PERCENT, value: '%' });
        i++;
      } else if (ch === '(') {
        tokens.push({ type: ExprTokenType.LPAREN, value: '(' });
        i++;
      } else if (ch === ')') {
        tokens.push({ type: ExprTokenType.RPAREN, value: ')' });
        i++;
      } else if (/\d/.test(ch) || (ch === '.' && i + 1 < expr.length && /\d/.test(expr[i + 1] || ''))) {
        let numStr = '';
        while (i < expr.length && (/\d/.test(expr[i] || '') || expr[i] === '.')) {
          numStr += expr[i];
          i++;
        }
        tokens.push({ type: ExprTokenType.NUMBER, value: numStr });
      } else {
        throw new LMONMacroError(`Unexpected character in expression: '${ch}'`, line, col);
      }
    }

    tokens.push({ type: ExprTokenType.EOF, value: '' });
    return tokens;
  }

  private current(): ExprToken {
    return this.tokens[this.pos] || { type: ExprTokenType.EOF, value: '' };
  }

  private advance(): ExprToken {
    const token = this.tokens[this.pos++];
    return token || { type: ExprTokenType.EOF, value: '' };
  }

  private expect(type: ExprTokenType, line: number, col: number): ExprToken {
    const token = this.current();
    if (token.type !== type) {
      throw new LMONMacroError(
        `Expected ${type} but got ${token.type}`,
        line,
        col,
      );
    }
    return this.advance();
  }

  private parseAddExpr(line: number, col: number): number {
    let left = this.parseMulExpr(line, col);

    while (
      this.current().type === ExprTokenType.PLUS ||
      this.current().type === ExprTokenType.MINUS
    ) {
      const op = this.advance();
      const right = this.parseMulExpr(line, col);
      left = op.type === ExprTokenType.PLUS ? left + right : left - right;
    }

    return left;
  }

  private parseMulExpr(line: number, col: number): number {
    let left = this.parseUnary(line, col);

    while (
      this.current().type === ExprTokenType.STAR ||
      this.current().type === ExprTokenType.SLASH ||
      this.current().type === ExprTokenType.PERCENT
    ) {
      const op = this.advance();
      const right = this.parseUnary(line, col);

      if (op.type === ExprTokenType.STAR) {
        left = left * right;
      } else if (op.type === ExprTokenType.SLASH) {
        if (right === 0) {
          throw new LMONMacroError('Division by zero in expression', line, col);
        }
        left = left / right;
      } else {
        if (right === 0) {
          throw new LMONMacroError('Modulo by zero in expression', line, col);
        }
        left = left % right;
      }
    }

    return left;
  }

  private parseUnary(line: number, col: number): number {
    if (this.current().type === ExprTokenType.MINUS) {
      this.advance();
      return -this.parseUnary(line, col);
    }
    return this.parsePrimary(line, col);
  }

  private parsePrimary(line: number, col: number): number {
    if (this.current().type === ExprTokenType.NUMBER) {
      const val = parseFloat(this.advance().value);
      return val;
    }

    if (this.current().type === ExprTokenType.LPAREN) {
      this.advance();
      const val = this.parseAddExpr(line, col);
      this.expect(ExprTokenType.RPAREN, line, col);
      return val;
    }

    throw new LMONMacroError(
      `Expected number or '(' in expression, got ${this.current().type}`,
      line,
      col,
    );
  }
}

export function expand(input: string, options?: ExpandOptions): string {
  const specMacros = buildSpecMacros();
  const context: MacroContext = new Map([...specMacros, ...(options?.initialContext ?? [])]);
  const strict = options?.strict ?? true;
  const maxDepth = options?.maxDepth ?? 16;

  const lines = input.split('\n');
  const outputLines: string[] = [];
  const trailingNewline = input.endsWith('\n');

  for (let lineNum = 1; lineNum <= lines.length; lineNum++) {
    const line = lines[lineNum - 1] || '';
    const defMatch = line.match(
      /^\s*%([a-zA-Z_][a-zA-Z0-9_]*)(?:\(([^)]*)\))?\s*=\s*"((?:[^"\\]|\\.)*)"\s*$/,
    );

    if (defMatch) {
      const name = defMatch[1] || '';
      const paramStr = defMatch[2] || '';
      const rawBody = defMatch[3] || '';

      const params =
        paramStr.length > 0
          ? paramStr
              .split(',')
              .map((p) => p.trim())
              .filter((p) => p.length > 0)
          : null;

      const body = decodeEscapes(rawBody);

      context.set(name, {
        body,
        params: params && params.length > 0 ? params : null,
        sourceLine: lineNum,
      });
    } else {
      const expanded = expandLine(
        line,
        context,
        strict,
        maxDepth,
        lineNum,
        0,
      );
      outputLines.push(expanded);
    }
  }

  let result = outputLines.join('\n');
  if (trailingNewline) {
    result += '\n';
  }
  return result;
}

function decodeEscapes(s: string): string {
  return s
    .replace(/\\"/g, '"')
    .replace(/\\\\/g, '\\')
    .replace(/\\n/g, '\n')
    .replace(/\\t/g, '\t');
}

function generateUUID(): string {
  const chars = '0123456789abcdef';
  let uuid = '';
  for (let i = 0; i < 36; i++) {
    if (i === 8 || i === 13 || i === 18 || i === 23) {
      uuid += '-';
    } else if (i === 14) {
      uuid += '4';
    } else if (i === 19) {
      uuid += chars[Math.floor(Math.random() * 16) & 0x3 | 0x8];
    } else {
      uuid += chars[Math.floor(Math.random() * 16)];
    }
  }
  return uuid;
}

function buildSpecMacros(): MacroContext {
  const now = new Date();

  const padZero = (n: number): string => String(n).padStart(2, '0');

  const dateStr: string = now.toISOString().split('T')[0] || '';
  const timeStr: string = `${padZero(now.getHours())}:${padZero(now.getMinutes())}:${padZero(now.getSeconds())}`;
  const datetimeStr: string = now.toISOString();

  const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  const dayStr: string = days[now.getDay()] || 'Sunday';

  const context: MacroContext = new Map();
  context.set('_DATE_STR', { body: dateStr, params: null, sourceLine: 0 });
  context.set('_TIMESTAMP', { body: String(Math.floor(now.getTime() / 1000)), params: null, sourceLine: 0 });
  context.set('_DATETIME_STR', { body: datetimeStr, params: null, sourceLine: 0 });
  context.set('_DAY_STR', { body: dayStr, params: null, sourceLine: 0 });
  context.set('_TIME_STR', { body: timeStr, params: null, sourceLine: 0 });
  context.set('_UUID', { body: generateUUID(), params: null, sourceLine: 0 });
  context.set('_ENV', { body: '', params: ['VAR'], sourceLine: 0 });

  return context;
}

function expandLine(
  line: string,
  context: MacroContext,
  strict: boolean,
  maxDepth: number,
  lineNum: number,
  depth: number,
): string {
  if (depth > maxDepth) {
    throw new LMONMacroError(
      'Macro expansion depth exceeded',
      lineNum,
      1,
    );
  }

  let result = '';
  let i = 0;

  while (i < line.length) {
    if (line[i] === '%') {
      if (i + 1 < line.length && line[i + 1] === '{') {
        const closeIdx = line.indexOf('}', i + 2);
        if (closeIdx === -1) {
          throw new LMONMacroError(
            'Unclosed expression macro %{...}',
            lineNum,
            i + 1,
          );
        }

        let exprStr = line.substring(i + 2, closeIdx);
        exprStr = expandLine(
          exprStr,
          context,
          strict,
          maxDepth,
          lineNum,
          depth + 1,
        );

        const evaluator = new ExprEvaluator();
        const exprResult = evaluator.evaluate(exprStr, lineNum, i + 1);
        const stringified = Number.isInteger(exprResult)
          ? String(Math.floor(exprResult))
          : String(exprResult);

        result += stringified;
        i = closeIdx + 1;
      } else {
        const nameMatch = line.substring(i + 1).match(/^([a-zA-Z_][a-zA-Z0-9_]*)/);
        if (!nameMatch) {
          result += '%';
          i++;
          continue;
        }

        const name = nameMatch[1] || '';
        let nextIdx = i + 1 + name.length;
        let args: string[] | null = null;

        if (nextIdx < line.length && line[nextIdx] === '(') {
          const closeParenIdx = line.indexOf(')', nextIdx);
          if (closeParenIdx === -1) {
            throw new LMONMacroError(
              `Unclosed parameter list for macro '${name}'`,
              lineNum,
              nextIdx + 1,
            );
          }

          const argStr = line.substring(nextIdx + 1, closeParenIdx);
          args = argStr.length > 0 ? argStr.split(',').map((a) => a.trim()) : [];
          nextIdx = closeParenIdx + 1;
        }

        const def = context.get(name);

        if (!def) {
          if (strict) {
            throw new LMONMacroError(
              `Undefined macro '${name}'`,
              lineNum,
              i + 1,
              name,
            );
          }
          result += `%${name}${args ? `(${args.join(',')})` : ''}`;
          i = nextIdx;
          continue;
        }

        if (def.params === null && args && args.length > 0) {
          throw new LMONMacroError(
            `Macro '${name}' takes no parameters, but ${args.length} provided`,
            lineNum,
            i + 1,
            name,
          );
        }

        if (def.params !== null && (!args || args.length !== def.params.length)) {
          const expected = def.params.length;
          const got = args?.length ?? 0;
          throw new LMONMacroError(
            `Macro '${name}' expects ${expected} arguments, got ${got}`,
            lineNum,
            i + 1,
            name,
          );
        }

        let substituted: string;

        if (name === '_ENV' && args && args.length > 0) {
          const envVarName = args[0] || '';
          const processEnv = (globalThis as any).process?.env;
          const envValue = (processEnv && typeof processEnv === 'object' && envVarName in processEnv) ? processEnv[envVarName] : '';
          substituted = envValue || '';
        } else {
          let body = def.body;

          if (def.params && args) {
            for (let j = 0; j < def.params.length; j++) {
              const placeholder = `{${def.params[j]}}`;
              const argVal = args[j] || '';
              body = body.replaceAll(placeholder, argVal);
            }
          }

          substituted = expandLine(
            body,
            context,
            strict,
            maxDepth,
            lineNum,
            depth + 1,
          );
        }

        result += substituted;
        i = nextIdx;
      }
    } else {
      result += line[i];
      i++;
    }
  }

  return result;
}
